Support batch PDB generation.

sonyps5201314 opened the pull request:
Added support for multi-select batch PDB generation. When multiple assemblies are selected, the user is prompted to choose a target folder, and files are generated as `{ShortName}.pdb` based on the assembly's short name.

siegfriedpammer commented on ILSpy/Commands/GeneratePdbContextMenuEntry.cs:
```diff
@@ -105,6 +116,109 @@ internal static void GeneratePdbForAssembly(LoadedAssembly assembly, LanguageSer
+						if (options.Progress != null)
+						{
+							options.Progress.Report(new DecompilationProgress {
+								Title = "Generating portable PDB...",
```
@sonyps5201314 if you want, you can add this. There are other locations where this same string is used, that might need updating as well.

sonyps5201314 replied:
https://github.com/icsharpcode/ILSpy/pull/3619/commits/71cfce3a486152db9e3139b98ab6c4806459e053 resolved.

siegfriedpammer commented on ILSpy/Commands/GeneratePdbContextMenuEntry.cs:
```diff
@@ -105,6 +116,109 @@ internal static void GeneratePdbForAssembly(LoadedAssembly assembly, LanguageSer
+
+		internal static void GeneratePdbForAssemblies(System.Collections.Generic.IEnumerable<LoadedAssembly> assemblies, LanguageService languageService, DockWorkspace dockWorkspace)
+		{
+			var assemblyArray = assemblies?.Where(a => a != null).ToArray();
```
```suggestion
			var assemblyArray = assemblies?.Where(a => a != null).ToArray() ?? [];
```

then you can omit the null checks later.

also I would assume that nobody passes null in here... at some point this needs nullable annotations

siegfriedpammer commented:
Looks good to me! Thank you very much  for contributing this feature!

sonyps5201314 commented on ILSpy/Commands/GeneratePdbContextMenuEntry.cs:
```diff
@@ -105,6 +116,109 @@ internal static void GeneratePdbForAssembly(LoadedAssembly assembly, LanguageSer
+							{
+								var decompiler = new CSharpDecompiler(file, assembly.GetAssemblyResolver(options.DecompilerSettings.AutoLoadAssemblyReferences), options.DecompilerSettings);
+								decompiler.CancellationToken = ct;
+								PortablePdbWriter.WritePdb(file, decompiler, options.DecompilerSettings, stream, progress: options.Progress);
```
https://github.com/icsharpcode/ILSpy/pull/3619/commits/ef52899b0bd7122f7bc3ec922d2bc46534e1521a resolved

siegfriedpammer commented on ILSpy/Commands/GeneratePdbContextMenuEntry.cs:
```diff
@@ -105,6 +116,109 @@ internal static void GeneratePdbForAssembly(LoadedAssembly assembly, LanguageSer
+			}
+
+			// Ask for target folder
+			using (var dlg = new System.Windows.Forms.FolderBrowserDialog())
```
Got it, thanks!

riQQ replied:
Alternatively, as of .NET 8, there's also an OpenFolderDialog in WPF:
https://learn.microsoft.com/en-us/dotnet/api/microsoft.win32.openfolderdialog?view=windowsdesktop-8.0

sonyps5201314 replied:
https://github.com/icsharpcode/ILSpy/pull/3619/commits/a973b2e71d1d606980d3cade0df8ced9a91d4901 resolved.

siegfriedpammer commented on ILSpy/Util/ShellHelper.cs:
```diff
@@ -0,0 +1,152 @@
+
+namespace ICSharpCode.ILSpy.Util
+{
+	static class ShellHelper
```
@sonyps5201314 why not just keep Process.Start("explorer.exe", ....)? We'd prefer to not introduce more P/Invoke dependencies.

sonyps5201314 commented:
@siegfriedpammer I believe I have addressed all the points you mentioned. The functionality of `GeneratePdbForAssembly` is now fully covered by `GeneratePdbForAssemblies`, so `GeneratePdbForAssembly` can be removed to simplify maintenance.

Additionally, I intentionally did not set an initial directory for `OpenFolderDialog` because I prefer to use the user's last selected directory. If we default to the target DLL's directory, there is a risk of overwriting an existing official PDB, which could break the software's integrity verification.

siegfriedpammer commented on ILSpy/Util/ShellHelper.cs:
```diff
@@ -0,0 +1,152 @@
+								}
+							}
+							if (itemPidl != IntPtr.Zero)
+								CoTaskMemFree(itemPidl);
```
I think, this needs try-finally too?

siegfriedpammer commented on ILSpy/Util/ShellHelper.cs:
```diff
@@ -0,0 +1,152 @@
+			}
+			catch
+			{
+				// ignore
```
what kind of exceptions are we ignoring here? "catch all" is considered bad pratice.

siegfriedpammer commented on ILSpy/TreeNodes/AssemblyTreeNode.cs:
```diff
@@ -690,15 +690,18 @@ public void Execute(TextViewContext context)
 		{
 			if (context.SelectedTreeNodes == null)
 				return;
+			var paths = new List<string>();
```
maybe it would make sense to use a HashSet<string> here? Note that this would need a `StringComparer.OrdinalIgnoreCase` on Windows :)

sonyps5201314 replied:
I don't think a change is strictly necessary here. If multiple instances of the same file appear in the Assembly List, it implies that the insertion logic (`AssemblyList.OpenAssembly`) isn't handling things correctly. The restriction should be enforced at that layer, rather than becoming a burden for internal calls.

 Given that `AssemblyList` defines:
 `readonly Dictionary<string, LoadedAssembly> byFilename = new Dictionary<string, LoadedAssembly>(StringComparer.OrdinalIgnoreCase);`
 we can infer that ILSpy doesn't suffer from this issue. However, I did notice a few days ago that `dnSpyEx` indeed has a problem with duplicate file entries.

 That being said, your point is valid. Since relying on external restrictions increases the cognitive load on the caller (requiring the user to pay special attention), I went ahead and handled it inside `OpenFolderAndSelectItems`.
https://github.com/icsharpcode/ILSpy/pull/3619/commits/752e6ed9c049e95923581a3d16d997d5e36e6be5 resolved.

sonyps5201314 commented on ILSpy/Util/ShellHelper.cs:
```diff
@@ -0,0 +1,152 @@
+using System.Runtime.InteropServices;
+using System.Linq;
+using System.Collections.Generic;
+
```
I didn't actually get this warning when I compiled, but I still added it as you suggested.
https://github.com/icsharpcode/ILSpy/pull/3619/commits/b631b55054dc90e2faf2f415c4e4c451202793a4 resolved.
